pkg/reconciler/reconciler.go

```diff
@@ -0,0 +1,171 @@
+	kubeClient := kubernetes.NewForConfigOrDie(cfg)
+	eventingClient := clientset.NewForConfigOrDie(cfg)
+	dynamicClient := dynamic.NewForConfigOrDie(cfg)
+
+	configMapWatcher := configmap.NewInformedWatcher(kubeClient, system.Namespace())
+
+	return Options{
+		KubeClientSet:     kubeClient,
+		DynamicClientSet:  dynamicClient,
+		EventingClientSet: eventingClient,
+		ConfigMapWatcher:  configMapWatcher,
+		Logger:            logger,
+		ResyncPeriod:      10 * time.Hour, // Based on controller-runtime default.
+		StopChannel:       stopCh,
+	}
+}
+
+// GetTrackerLease returns a multiple of the resync period to use as the
+// duration for tracker leases. This attempts to ensure that resyncs happen to
+// refresh leases frequently enough that we don't miss updates to tracked
+// objects.
+func (o Options) GetTrackerLease() time.Duration {
+	return o.ResyncPeriod * 3
+}
+
+// Base implements the core controller logic, given a Reconciler.
+type Base struct {
+	// KubeClientSet allows us to talk to the k8s for core APIs
+	KubeClientSet kubernetes.Interface
+
+	// EventingClientSet allows us to configure Eventing objects
+	EventingClientSet clientset.Interface
+
+	// DynamicClientSet allows us to configure pluggable Build objects
+	DynamicClientSet dynamic.Interface
+
+	// ConfigMapWatcher allows us to watch for ConfigMap changes.
+	ConfigMapWatcher configmap.Watcher
+
+	// Recorder is an event recorder for recording Event resources to the
+	// Kubernetes API.
+	Recorder record.EventRecorder
+
+	// StatsReporter reports reconciler's metrics.
+	//StatsReporter StatsReporter
+
+	// Sugared logger is easier to use but is not as performant as the
+	// raw logger. In performance critical paths, call logger.Desugar()
+	// and use the returned raw logger instead. In addition to the
+	// performance benefits, raw logger also preserves type-safety at
+	// the expense of slightly greater verbosity.
+	Logger *zap.SugaredLogger
+}
+
+// NewBase instantiates a new instance of Base implementing
+// the common & boilerplate code between our reconcilers.
+func NewBase(opt Options, controllerAgentName string) *Base {
+	// Enrich the logs with controller name
+	logger := opt.Logger.Named(controllerAgentName).With(zap.String(logkey.ControllerType, controllerAgentName))
+
+	recorder := opt.Recorder
+	if recorder == nil {
+		// Create event broadcaster
+		logger.Debug("Creating event broadcaster")
+		eventBroadcaster := record.NewBroadcaster()
+		watches := []watch.Interface{
+			eventBroadcaster.StartLogging(logger.Named("event-broadcaster").Infof),
+			eventBroadcaster.StartRecordingToSink(
+				&typedcorev1.EventSinkImpl{Interface: opt.KubeClientSet.CoreV1().Events("")}),
+		}
+		recorder = eventBroadcaster.NewRecorder(
+			scheme.Scheme, corev1.EventSource{Component: controllerAgentName})
+		go func() {
+			<-opt.StopChannel
+			for _, w := range watches {
+				w.Stop()
+			}
+		}()
+	}
+
+	//statsReporter := opt.StatsReporter
```

Comment:
TODO? Maybe make an issue to track this work?

Reply:
Tracking here: https://github.com/knative/eventing/issues/734#issuecomment-484280270